File: Data Structures & Algorithms/course-schedule/submission-2.java
class Solution {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        int[] indegree = new int[numCourses];
        List<List<Integer>> adj = new ArrayList<>();
        for(int i  = 0; i < numCourses; i++){
            adj.add(new ArrayList<>());
        }
        for(int[] pair : prerequisites) {
            int course = pair[0];
            int requirement = pair[1];
            indegree[course]++;
            adj.get(requirement).add(course);
        }
        Queue<Integer> q = new LinkedList<>();
        for(int i = 0; i < numCourses; i++){
            if (indegree[i] == 0) q.add(i);
        }
        int finished = 0;
        while(!q.isEmpty()){
            int node = q.poll();
            finished++;
            for(int nei : adj.get(node)){
                indegree[nei]--;
                if (indegree[nei] == 0) q.add(nei);
            }
        }
        return finished == numCourses;
    }
}

File: Data Structures & Algorithms/course-schedule-ii/submission-2.py
class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        res = []
        indegree = [0] * numCourses
        adj = defaultdict(list)
        finished = 0
        # 1. Traverse and check indegree
        for course, prereq in prerequisites:
            adj[prereq].append(course)
            indegree[course] += 1

        # 2. add indegree = 0 to queue
        q = deque()
        for i in range(numCourses):
            if indegree[i] == 0:
                q.append(i)

        # 3. traverse and reduce indegree.
        #   if indegree is 0, add it to queue
        while q:
            node = q.popleft()
            res.append(node)
            finished += 1
            for nei in adj[node]:
                indegree[nei] -= 1
                if indegree[nei] == 0:
                    q.append(nei)

        return res if finished == numCourses else []